File: problems/0003_longest-substring-without-repeating-characters/map/solution.rb
#!/usr/bin/ruby

# @param {String} s
# @return {Integer}
def length_of_longest_substring(s)
    m = Hash.new(256)
    max, start, e = [0, 0, 0]
    while e < s.length
        c = s[e]
        ok, idx = [m.has_key?(c), m[c]]
        m[c] = e
        e += 1
        if not ok or idx < start
            max = [max, e - start].max
        else
            start = idx + 1
        end
    end
    return max
end

File: problems/0003_longest-substring-without-repeating-characters/map/Solution.php
<?php
class Solution {
    /**
     * @param String $s
     * @return Integer
     */
    function lengthOfLongestSubstring($s) {
        $map = [];
        $max = 0;
        $start = 0;
        $l = strlen($s);
        for ($end = 0; $end < $l;) {
            $c = $s[$end];
            $idx = $map[$c];
            $map[$c] = $end++;
            if (!isset($idx) || $idx < $start)
                $max = max($max, $end - $start);
            else
                $start = $idx + 1;
        }
        return $max;
    }
}
